src/curvefit/uncertainty/predictive_validity.py

```diff
@@ -0,0 +1,139 @@
+        are absolute residuals. If the evaluation space is in linear space, then `theta`
+        is set to `1.` so that they are relative residuals to the prediction magnitude.
+    - `group_residuals (Dict[str: curvefit.uncertainty.residuals.Residual]): residual objects
+        for each group in the dataset
+    - `group_records (Dict[str: List[curvefit.solvers.solver.Solver]]): a list of solvers
+        at each time point in the predictive validity analysis for a particular group
+
+    ## Methods
+
+    ### `_make_group_info`
+    Creates a [ResidualInfo](ResidualInfo.md) object based on the data passed in and
+    for a particular group.
+
+    - `data (curvefit.core.data.DataSpecs)`: data specifications
+    - `group_name (str)`: the name of the group
+
+    ### `_run_group_pv`
+    Runs the predictive validity analysis for only one group using the `data`, `model`
+    and `solver` from the `run_predictive_validity()` function.
+
+    ### `run_predictive_validity`
+    Runs predictive validity for all groups in the data.
+
+    - `data (curvefit.core.data.Data)`: data and specifications for the whole analysis
+    - `group (str)`: name of the group
+    - `model (curvefit.models.base.Model)`: a model object that may be copied depending
+        on whether or not in debug mode
+    - `solver (curvefit.solvers.solver.Solver)`: a solver used to fit the model
+
+    ### `get_residual_data`
+    Return the out of sample residuals for all groups from the data argument to
+    `run_predictive_validity`. Important input to the [`_ResidualModel`](_ResidualModel.md) for
+    eventually creating uncertainty.
+
+    {end_markdown PredictiveValidity}
+    """
+    def __init__(self, evaluation_space, debug_mode=False):
+
+        assert callable(evaluation_space)
+        assert type(debug_mode) == bool
+
+        self.evaluation_space = evaluation_space
+        self.debug_mode = debug_mode
+
+        if self.evaluation_space.__name__.startswith('ln'):
+            self.theta = 0.
+        else:
+            self.theta = 1.
+
+        self.group_residuals = dict()
+        self.group_records = dict()  # not filled if not in debug_mode
+
+    @staticmethod
+    def _make_group_info(data, group_name):
+        df, specs = data._get_df(group=group_name, copy=False, return_specs=True)
+
+        return ResidualInfo(
+            group_name=group_name,
+            times=df[specs.col_t].values,
+            obs=df[specs.col_obs].values
+        )
+
+    def _run_group_pv(self, data, group, model, solver):
+
+        if self.debug_mode:
+            self.group_records[group] = []
+
+        for i, t in enumerate(self.group_residuals[group].residual_info.times):
+
+            if self.debug_mode:
+                model = model.clone()
+                solver = solver.clone()
+                solver.set_model_instance(model)
+
+            df = data._get_df(group=group).query(f"{data.data_specs.col_t} <= {t}")
+            solver.fit(data=df)
+
```

Comment:
I think we may want to `set_model_instance` for non-debug mode too ? also just a thought (the current code looks very clean and I do not wish to change it), but we maybe can fit with all data first, save the `data_inputs`, and then use only a subset of rows of matrices(i.e. not using all the data). maybe I should add a field `n_times` in `DataInputs` to say how many rows we want to use. This could reduce dataframe access, but not as clean or human readable.